localstack/services/apigateway/helpers.py

```diff
@@ -706,25 +708,34 @@ def get_resource_for_path(path: str, path_map: Dict[str, Dict]) -> Optional[Tupl
 
     # if there are no matches, it's not worth to proceed, bail here!
     if not matches:
-        return None
+        return None, None
 
     # so we have matches and perhaps more than one, e.g
     # /{proxy+} and /api/{proxy+} for inputs like /api/foo/bar
     # /foo/{param1}/baz and /foo/{param1}/{param2} for inputs like /for/bar/baz
     if len(matches) > 1:
-        # check if we have an exact match (exact matches take precedence)
-        for match in matches:
-            if match[0] == path:
-                return match
-
-        # not an exact match but parameters can fit in
+        filtered_matches = []
         for match in matches:
```

Comment:
Ok I see what happened here, we used to do this (https://github.com/localstack/localstack/blob/0c93517c4fa5372f0236b70d5f4f7b5c55ad3164/localstack/services/apigateway/helpers.py#L1340) which since has not being used, but if we are moving this logic `get_resource_for_path` we should account for the uppercase mismatches, and the `X-AMAZON-APIGATEWAY-ANY-METHOD` situation.


Also remove the `get_target_resource_method` since is not used anymore and probably the cause of this gap.

```
def get_target_resource_method(invocation_context: ApiInvocationContext) -> Optional[Dict]:
    """Look up and return the API GW resource method for the given invocation context."""
    _, resource = get_target_resource_details(invocation_context)
    if not resource:
        return None
    methods = resource.get("resourceMethods") or {}
    method_name = invocation_context.method.upper()
    return (
        methods.get(method_name)
        or methods.get("ANY")
        or methods.get("X-AMAZON-APIGATEWAY-ANY-METHOD")
    )
```

Reply:
I don't think this is the same issue. 
`get_target_resource_method ` is still used in `get_target_resource_method_or_route` so we shouldn't remove it.

It does handle the method match afterwards, and we can still use it that way. 

It's just that if there is an exact match, we would return the resource, then try to get the method and there wouldn't be one. So we would just raise 404. But we need to be conscious of what methods a resource implements on the matching already. I don't think we need to change the logic yet, it works properly like that even if there is some duplication. 

Now we match the route, then get the method for it. Before we would match only on the path, then try to get the method even if it wasn't implemented. 

Also, we don't need to take care of `X-AMAZON-APIGATEWAY-ANY-METHOD` anymore since this has been fixed in the OpenAPI importer, it's now properly converted to `ANY` (that's what AWS stores internally as well). I can change the logic to take care of the uppercase mismatch but I don't think it's an issue either, because the `InvocationContext` method comes from the routing framework and this is always uppercase.

Reply:
After checking, it seems `find_best_candidate` has the proper logic there, but it's only used for `find_matching_http_route`

Reply:
Sounds good, the only hit comes from `ext` and that mislead me. The logic to take in account the method also makes sense.

Yes the `find_best_candidate` was part of a refactor but I think the rules of precedence are a bit different.